### ISSUE TITLE
OperatorDependecyTree is bugged. AMIE is faster than we thought.

### ISSUE THREAD

**Issue by falcaopetri** [lajus/amie#41]
> **Disclaimer**: I'm making a hard claim here, so I'm still a little suspicious about all of this.

**TL;DR.** When using any mining assistant (excluding the base `MiningAssistant`), there might be (and **there actually is**) duplicated invocations of the mining operators. Each duplicated invocation overrides the previous. As they produce the same set of rules, this does not change the output but causes **overhead in execution time**.

## Contextualization

I got very interested in `InjectiveMappingsAssistant` since I discovered it. Unfortunately, it took almost 10x more time to execute my experiment. Remember that I'm mining up-to-4-atom rules.

This seemed very strange since the overhead of injectiveness should not be so huge.

## Hard to debug
I've started debugging some days ago and even used JProfiler for profiling AMIE. There were many weird things occurring when comparing Lazy and Injective profiling results, but it took me a long time to understand what was going on.

Long story short, I ended up at `MiningAssistant.applyMiningOperators`, and then at `MiningAssistant.computeOperatorHierarchy`. Things started making sense when I printed out the computed `MiningAssistant.miningOperators` for `LazyMiningAssistant`:

```java
public void amie.mining.assistant.MiningAssistant.getDanglingAtoms(amie.rules.Rule,double,java.util.Collection)
public void amie.mining.assistant.DefaultMiningAssistant.getClosingAtoms(amie.rules.Rule,double,java.util.Collection)
public void amie.mining.assistant.DefaultMiningAssistant.getDanglingAtoms(amie.rules.Rule,double,java.util.Collection)
public void amie.mining.assistant.MiningAssistant.getInstantiatedAtoms(amie.rules.Rule,double,java.util.Collection,java.util.Collection)
public void amie.mining.assistant.MiningAssistant.getClosingAtoms(amie.rules.Rule,double,java.util.Collection)
public void amie.mining.assistant.DefaultMiningAssistant.getTypeSpecializedAtoms(amie.rules.Rule,double,java.util.Collection)
```

## Hard to spot

Read again the name of the methods considered as mining operators. How was it not noted that during `LazyMiningAssistant`, e.g., both `MiningAssistant.getClosingAtoms` and `DefaultMiningAssistant.getClosingAtoms` were being called? Well... because they weren't.

Due to polymorphism (I guess), both `getClosingAtoms` ended up calling the most specific version (in this case, the `DefaultMiningAssistant.getClosingAtoms`). Therefore, the same operator was called twice. Or three times, in the case of `InjectiveMiningAssistant`.

Since the operators have the same name, `MiningAssistant.applyMiningOperators` returned the "expected" output.

## Fix

See the associated PR for fix proposal and some basic benchmarks.

> (Autolinking PR failed. Here it is: #42).

Another possibility I considered was to make `MiningAssistant.miningOperators` a `Set`, but not allowing duplicates in `OperatorDependencyTree` sounded like a better way.

**Comment by lajus** [comment]
Hi Antonio,

Sorry for the late reply, I am quite busy at the moment.

I am not very familiar with this part of the code as it is Luis' code and the dynamic loading in general. @lgalarra can you confirm the problem and verify the proposed fix ?

Thanks for spotting this,
Jonathan

**Comment by lgalarra** [comment]
I'll look into it in the upcoming days.

**Comment by lgalarra** [comment]
The proposed fix seems correct to me. The mining operators are explored from the more specific to the more general classes, that is, if the most specific assistant defines all the operators, we will not consider the operators from the upper classes. If those methods need to be invoked, it is the subclass responsibility to do it via super(). I also agree that -verbose should output the operators that are applicable!

**Comment by lgalarra** [comment]
I have implemented the output of the mining operators and committed it to the master branch.
